### ISSUE TITLE
How to reproduce this LoRD method?

### ISSUE THREAD

**Issue by Dream-Camellia** [liangzid/LoRD-MEA#3]
@liangzid Hello author, I would like to replicate the work of this project. I think it is very meaningful and innovative, but I currently do not understand how to reproduce the code from the readme. Could you please provide guidance, such as which parts to run first and which to run afterwards, or what the general content of the code is?
Thank you for your guidance.

**Comment by liangzid** [comment]
Hi Dream-Camellia,  

I sincerely apologize for the delayed response. Your issue notification was unfortunately buried among hundreds of GitHub issue emails from other people's repositories I watched, so I missed it until now.  

Regarding your feedback, I’ll be updating the ReadMe file shortly with more detailed instructions after the NIPS ddl.
To better assist you, could you clarify whether your focus is on:  
1. **Using LoRD as a method** (i.e., applying it to your own use case), or  
2. **Reproducing our experimental results** from the paper?  

For the latter, you can refer to the `Evaluation` section in the current ReadMe and follow these steps:  

### Steps to Reproduce Experimental Results  
1. **Environment Setup**  
   Install dependencies via:  
   ```bash
   pip install -r re.txt
   ```  

2. **Model Training**  
   Run the desired script, e.g., for WMT16's cs-en dataset:  
   ```bash
   bash scripts/2.5.wmt_varying_methods.sh
   ```  

3. **Model Evaluation**  
   Use `wmt_process.py` to evaluate trained models:  
   ```bash
   python wmt_process.py
   ```  
   - The script includes functions for various experiments (varying model sizes, query numbers, etc.).  
   - To focus on a specific task, comment out other functions and configure the `main()` function with your model’s path:  
     ```python
     infer_wmt(
         modelname,         # Path to your trained model
         task_name,        # Task identifier (e.g., "cs-en")
         res_pth,         # Output path for results
         test_set_take_num, # Number of test samples
         mnt=16,          # Tokens generated per query
         base_model_name=None  # Pretrained model path (if using LoRA)
     )
     ```  

This should serve as a starting point for reproducing one of the key results.  

If you encounter any issues (unaddressed questions, roadblocks in the process, or code errors), please feel free to post them here. I’m happy to help!

**Comment by Dream-Camellia** [comment]
Dear author！Thank you for sending me the message! @liangzid 
Thank you very much for your reply to me. I am very interested in your research in this area. I want to reproduce it in the fastest way or rather, I want to use this LoRD as a method. However, after reading the paper, I still have some doubts about the code and need your guidance so that I can understand the code work of this LoRD as soon as possible.
Could you guide me? How can I carry out the theft work of this LoRD, to use the small model to steal part of the capabilities of the large model? How can I implement this core part of the functional algorithm, or which function's code is the core py code of the reproduction paper? I will carefully study it.

I really want to learn this LoRD method. I hope you can give me this opportunity. I would be very grateful. @liangzid

**Comment by liangzid** [comment]
Hi, Dream-Camellia~

Let me give a brief introduction to this source code.

1. Regarding the loss function we used for stealing, it is in `train_pod2.py`'s `one-period` function. Specifically, we use the `"LoRD-VI"` in our experiments, with the implementation procedure of:
```python
              logits11 = lm(idxs11).logits[:, :-1, :]
              logits11 = F.log_softmax(logits11, dim=-1)
              logits11 = logits11[torch.arange(bs).unsqueeze(1),
                                  torch.arange(sqlen-1).unsqueeze(0),
                                  idxs11[:, 1:sqlen]]
  
              logits12 = lm(idxs12).logits[:, :-1, :]
              logits12 = F.log_softmax(logits12, dim=-1)
              logits12 = logits12[torch.arange(bs).unsqueeze(1),
                                  torch.arange(sqlen-1).unsqueeze(0),
                                  idxs12[:, 1:sqlen]]
  
              if args.is_black_box == 0:
                    term3 = \
                        (vic_logits2[:, :, 0]-logits2_cons)
                else:
                    # term3 = old_logits2 - logits2_cons
                    term3 = - logits2_cons
                term3 = torch.mean(log_clip(term3))

                los2=-1*torch.mean(logits2_cons)
                loss11=-1*torch.mean(logits11)
                loss12=1*torch.mean(logits12)

                loss = los2+loss11+2*loss12
```
where `logits11` and `logits12` respectively denote the likelihood of positive and negative responses generated by the local model, i.e., $y_+$ and $y_-$.

2. The training procedure. Our training procedure is provided in `train_pod2.py`, the `train_pod` function. While the whole function is complicated, the core part is from line 608 to line 638, as shown below:
```python
                if p12 > p11:
                # if delta11<delta12:
                    print("SWAP.")
                    p_i_11_ls[i] = pidx12.squeeze(0).to("cpu")
                    p_i_12_ls[i] = pidx11.squeeze(0).to("cpu")

                    temppp = p_logits_11_ls[i]
                    p_logits_11_ls[i] = p_logits_12_ls[i]
                    p_logits_12_ls[i] = temppp

                    temppp = p_m_11_ls[i]
                    p_m_11_ls[i] = p_m_12_ls[i]
                    p_m_12_ls[i] = temppp

                    temppp = delta12
                    delta12=delta11
                    delta11=temppp
                # if max(p11, p12) < tau1 or abs(p11-p12)<0.01:
                # if delta11 < tau1:
                if max(p11, p12) < tau1 and delta11<tau_delta:
                    print("RED:->BUT still use the VIC's labels.")
                    # print(f"shape of 11: {p_i_11_ls.shape}")
                    # print(f"shape of 2: {idx2ls.shape}")
                    # print(f"shape of 12: {p_i_12_ls.shape}")

                    p_i_11_ls[i] = p_i2ls[i]
                    p_m_11_ls[i] = pmask2s[i]
                    p_logits_11_ls[i] = p_logits2ls[i]

                if min(p11, p12) < tau2:
                   period_break = 0
```
3. Run Training. `lord_train.py` is the main entry of our training code. Here is an example of how to execute this file with parameters.
```sh
echo "HOME: ${HOME}"
export python=${HOME}/anaconda3/envs/align/bin/python3
# export CUDA_VISIBLE_DEVICES="5,6,7"
# export CUDA_VISIBLE_DEVICES="0,1,2"
export CUDA_VISIBLE_DEVICES="1"
export TORCH_USE_CUDA_DSA="1"
export root_dir="${HOME}/alignmentExtraction/"
export POD_save_dir="${root_dir}/wmt16_ckpts/"
export from_path="meta-llama/Meta-Llama-3-8B-Instruct"
export TRAIN_NUMS=(16)
# export TRAIN_NUMS=(64 128 256)
export train_times=(1 2 3 4 5)
# export train_times=(1)
export msl=256
export task_ls=("cs-en" "de-en" "fi-en") # three tasks in WMT16
export train_taskls=("LoRD-VI" "vanilla")


export is_black_box=1
export use_lora=1

# export epoch=3
# export period=3

export epoch=2
export period=1

export sub_set_num=1
export sub_stage_num=256
export max_new_tokens=64
export infer_batch_size=1
export batch_size=1

export beta=-1
export temperature=-1

export use_old_logits=1
export use_vic_logits=1
export use_kld=0
export use_entropy=0

# export tau1=0.85
export tau1=0.80
export tau2=0.85

for train_num in ${TRAIN_NUMS[*]}
do
    for train_time in ${train_times[*]}
    do
	for task in ${task_ls[*]}
	do
	    for train_task in ${train_taskls[*]}
	    do
		echo "====================================================="
		echo "+++++++train_num: ${train_num}+++++++"
		echo "+++++++train_time: ${train_time}+++++++"
		echo "+++++++task: ${task}+++++++"
		echo "+++++++train_task: ${train_task}+++++++"
		echo "====================================================="

		export save_path="${POD_save_dir}WMTTT0519${task}${train_num}${train_time}${train_task}"

		$python ${root_dir}lord_train.py\
		    --use_lora=$use_lora \
		    --from_path=$from_path \
		    --is_black_box=$is_black_box \
		    --sub_set_num=$sub_set_num \
		    --sub_stage_num=$sub_stage_num\
		    --infer_batch_size=$infer_batch_size\
		    --tau1=$tau1 \
		    --tau2=$tau2 \
		    --task=$train_task \
		    --device="cuda" \
		    --epoch=$epoch \
		    --period_num=$period \
		    --acc_step=1 \
		    --log_step=50 \
		    --train_num=$train_num \
		    --max_new_tokens=$max_new_tokens \
		    --LR="3e-5" \
		    --save_step=$sub_stage_num \
		    --beta=$beta \
		    --temperature=$temperature \
		    --batch_size=$batch_size \
		    --use_old_logits=$use_old_logits\
		    --use_vic_logits=$use_vic_logits\
		    --use_kld=$use_kld\
		    --max_length=$msl \
		    --dataset_task=$task \
		    --save_path=$save_path
		echo "DONE FOR ONE TRAIN NUMBERS...."
	    done
	done
    done
done


$python ${root_dir}wmt_process.py

```

The above introduction could improve the understanding of our source code. The other details (e.g., how to use this source code) are provided in my previous response. You can also take the readme file as a reference. Nonetheless, the source code remains difficult to use directly. I will provide a clean version of the code to help you and others to use LoRD.

Thank you for your reply. If you have any other concerns, please let me know :)

**Comment by Dream-Camellia** [comment]
Hello author, I have some questions about the LoRD method that I'd like to ask you.
@liangzid 
1. When conducting LoRD training, after each query, the corresponding local model will generate two responses. What are the main differences between these two responses? Or is it just that two responses were generated by a normal query and then calculated separately? Does this response refer to the output corresponding to the query of the model?
2. After conducting a query and completing the replacement of this y+t-1, based on the calculated loss, is it necessary to fine-tune the parameters of the local model? After the modification is completed, conduct a new query again? Can it be understood that each time a query is conducted, the local model needs to be updated based on this loss?
3. I carefully read this code and feel that it's a bit difficult to reproduce it because this project is rather complex. Where is the corresponding code module for the part of the LoRD method sequence number ⑤? I don't particularly understand the process of this part.

**Comment by liangzid** [comment]
Dear Dearm-Camellia,

Thanks for your feedback. Here are the responses:

1. `What are the main differences between these two responses?`: They are randomly sampled from the same token distribution $P(\cdot|x)$. This strategy is the same as the multiple action sampling in LLM's RL training.
2. `Does this response refer to the output corresponding to the query of the model?`: Yes.
3. `After conducting a query and completing the replacement of this y+t-1, based on the calculated loss, is it necessary to fine-tune the parameters of the local model?`: Yes, it is.
4. `After the modification is completed, conduct a new query again?`: In our experiments, we use a fixed query dataset which is divided into several mini-batch. After the parameter update on the previous batch, we will use the queries in the next batch.
5. ` Can it be understood that each time a query is conducted, the local model needs to be updated based on this loss?`: I think this is okay.
6. `this project is rather complex`: Yeah, it is really a dirty implementation with many noises.
7. `Where is the corresponding code module for the part of the LoRD method sequence number ⑤?`: Step 5 is about updating model parameters based on the calculated loss. It is in `train_pod2.py`'s [line1042-1046](https://github.com/liangzid/LoRD-MEA/blob/2be8bd25ef035eee300cb4693a9608ad55516e04/train_pod2.py#L1042)

I think the above responses can address your problems, and expect your further feedback.

BTW, regarding the item 3 you mentioned, I suggest you to take a look at the pseudocode of LoRD (page 24 in the paper and also shown below), which I think is more elegant. 

![Image](https://github.com/user-attachments/assets/a25baff0-d736-416f-8df5-aaa9a127130b)

P.S. in line 11 there seems a typo. `\mathcal{D}_0^+` should be the response set of the victim model's under queries.

Please feel free to further discuss this study~

**Comment by Dream-Camellia** [comment]
Hello author, thank you for answering my question. After reading your answer, I feel much clearer.
@liangzid 
1、Does this mean that initially there is a query set, and then a query is selected from it. For the local model, two responses are randomly generated, respectively as yt+ and yt-, and then participate in the calculation of the forward direction and loss later. After the calculation, backpropagation is performed to update this local model? After all the queries in this query set have been completed, can another query set be replaced?
2. However, for the following part of the pseudo-code (lines 11 to 15), I don't quite understand its meaning. Could you please explain it to me? Is it used to initialize the positive and negative sample sets?Why are the initialization processes different？

![Image](https://github.com/user-attachments/assets/5205bb9d-7fad-487f-944e-f9965fc3d7fa)

**Comment by liangzid** [comment]
Hi Dream-Camellia,

Thanks for your feedback. I think now I get the point of the problem. The main problem is that some points in Figure 2 and Algorithm 1 are incorrect, which are not aligned with our methodology description. 

I would first provide the response for your questions, and then show the incorrect points.

## Responses

This is what I think for your questions:
1. `Does this mean ... this local model?`: Yes. I think this is correct.
2. `After all the queries in this query set have been completed, can another query set be replaced?`: Yes. But we have not done that. I use the **same** query set and the **same** responses obtained from the victim model in our model extraction. This setting is totally the same as current supervised-based stealing. For more powerful stealing, I think there can be some design in the query set.
3. `Explanation of line 10-15`: Yes, this is a special step to obtain mini-batches $D_t^+$ and $D_t^-$ when t=0. 
   1. when the training step t>0: we use $P_{\theta_{t-1}}$ to generate two responses marked as $y_{t-1}^+$ and $y_{t-1}^-$ (see line 20), and use them in the training procedure of training step $t+1$;
   2. when the training step t=0: this is a special step because  $P_{\theta_{0}}$ is already the initial local model, and there is no $P_{\theta_{0-1}}$. So we use the generated responses of the initial local model as $y_0^-$ and the response generated from the victim model as $y_0^+$.

I think it might be confusing to understand why LoRD is designed to *use **previous** step's parameters of the model to generate samples for **next** step's training*. This is because we need a reward signal to discriminate positive and negative generated responses merely based on the victim model's response. And in this paper, we use the consistency of model optimization direction and the generated sample's optimization direction as the signal (which is represented by the value of $\Delta$). 
Figure 3 intuitively illustrates this idea:

![Image](https://github.com/user-attachments/assets/07910ee5-9041-4dab-b85d-4a3345f4498e)

## Errors in the Paper

This is a new version of the algorithm:

![Image](https://github.com/user-attachments/assets/9aecfff0-4222-4032-801b-bd129adf0596)

> Where is the corresponding code module for the part of the LoRD method sequence number ⑤? I don't particularly understand the process of this part.

Now I think the confusion is that in Figure 2 we incorrectly use the symbol $\theta_{t-1}$ in step 5.

![Image](https://github.com/user-attachments/assets/48a70024-8588-4e9a-a3e3-18f3799d8d5c)

Besides, there is also an error in the order of $\tau_1$ and $\tau_2$ in the main text: 

![Image](https://github.com/user-attachments/assets/d1f555c4-b072-41cc-80b0-8b5cb620aebf)

I reread the methodology part and recheck all digrams and the algorithm for their correctness, but I am not sure if there exist further errors. I sincerely appreciate your comments, and hope your further discussion of this study : ) Also, a revision of the paper will be released recently.

**Comment by Dream-Camellia** [comment]
Thank you very much for taking the time to answer my question.
@liangzid 
I'm very sorry. I just experienced the traditional Chinese festival, the Dragon Boat Festival, and didn't reply to your message in time.
I have gained a preliminary understanding of my problem. I think what I need to do now is to run this code once, and then I should be able to deeply understand this core idea. There should be many difficulties in the middle. I expect to use this method to do some new things because it is very good and innovative.
1、If I want to make the simplest use of this algorithm, in which code section should I focus, such as the loss function, backpropagation, and so on?
2、For this τ1 and τ2, how should this threshold define this value?
Because I think there might be some problems in reproduction for different experimental environments.
